yellow-paper/docs/contract-deployment/instances.md

```diff
@@ -17,102 +17,185 @@ The structure of a contract instance is defined as:
+
+The Deployment Nullifier is defined as the address of the contract being deployed. Note that it later gets [siloed](../circuits/private-kernel-tail.md#siloing-values) using the `ContractInstanceDeployer` address by the Kernel Circuit, so this nullifier is effectively the hash of the deployed contract address and the `ContractInstanceDeployer` address.
+
+Only in this state public function calls are valid. The Public Kernel Circuit validates that the Deployment Nullifier has been emitted by the `ContractInstanceDeployer` as part of its checks. Note that this requires hardcoding the address of an application-level contract in a protocol circuit.
+
+## Constructors
 
-Constructors may be either private or public functions. Contracts with private constructors can be privately or publicly deployed, contracts with public constructors can only be publicly deployed.
+Contract constructors are not enshrined in the protocol, but handled at the application circuit level. Constructors are methods used for initializing a contract, either private or public, and a contract may have more than a single constructor. A contract must ensure the following requirements are met:
 
-<!-- FIXME: the contract class definition says the `constructor_function` is always of type `PrivateFunction`. -->
+- A contract may be initialized at most once
+- A contract must be initialized using the method and arguments defined in its address preimage
+- All functions that depend on contract initialization cannot be invoked until the contract is initialized
 
-Removing constructors from the protocol itself simplifies the kernel circuit. Separating initialization from public deployment also enables developers to implement private deployments, since a private deployment is equivalent to just invoking the constructor function at a given address.
+These checks are embedded in the application circuits themselves. The constructor emits an Initialization Nullifier when it is invoked, which prevents it from being called more than once. The constructor code must also check that its own selector and the arguments for the call match the ones in the address preimage, which are supplied via an oracle call.
 
-## Public Deployment
+Private functions in the contract should require a merkle membership proof for the Initialization Nullifier, to prevent them from being called before the constructor is invoked. Nevertheless, a contract may choose to allow some functions to be called before initialization, such as in the case of [Diversified and Stealth account contracts](../addresses-and-keys/diversified-and-stealth.md). Public functions, on the other hand, should not need this check, since the Public Kernel Circuit guarantees that the contract is Publicly Deployed, and a contract cannot be Publicly Deployed without being Initialized. Note that every Contract Class needs to have at least one constructor if an instance of it is intended to be Publicly Deployed.
```

Comment:
No, because the initialization hash is emitted from the contract itself, not from the `Deployer`. 

Still, we can remove the restriction that "a contract needs to be initialized in order to be publicly deployed" as suggested near the end, which simplifies things (for us).